cosmwasm/enclaves/shared/contract-engine/src/contract_validation.rs

```diff
@@ -584,27 +645,160 @@ fn verify_callback_sig_impl(
+                    channel == packet.source_channel
+                    && sequence == packet.sequence
+                    && ack == String::from_utf8_lossy( acknowledgement)
+                    && success != is_transfer_ack_error(acknowledgement),
+                IBCLifecycleComplete::IBCLifecycleComplete(IBCLifecycleCompleteOptions::IBCTimeout { .. }) => false,
+            }
+                },
+            _ => false,
+
+            }
+        },
+        CosmosSdkMsg::MsgTimeout { packet, signer,.. } => {
+            match handle_type {
+                HandleType::HANDLE_TYPE_IBC_PACKET_TIMEOUT => {
+                    let send_msg_timeout_msg = serde_json::from_slice::<IBCPacketTimeoutMsg>(&sent_msg.msg);
+                    if send_msg_timeout_msg.is_err(){
+                        trace!("get_verified_msg HANDLE_TYPE_IBC_PACKET_TIMEOUT: sent_msg.msg cannot be parsed as IBCPacketTimeoutMsg: {:?} Error: {:?}", String::from_utf8_lossy(&sent_msg.msg), send_msg_timeout_msg.err());
+                        return false;
+                    }
+                    let sent_msg_timeout_msg= send_msg_timeout_msg.unwrap();
+
+                    sent_msg_timeout_msg.packet.src.channel_id == packet.source_channel &&
+                    sent_msg_timeout_msg.packet.src.port_id == packet.source_port &&
+                    sent_msg_timeout_msg.packet.dest.channel_id == packet.destination_channel &&
+                    sent_msg_timeout_msg.packet.dest.port_id == packet.destination_port &&
+                    sent_msg_timeout_msg.packet.sequence == packet.sequence &&
+                    sent_msg_timeout_msg.packet.data.0 == packet.data &&
+                    sent_msg_timeout_msg.relayer == *signer
+                },
+                HandleType::HANDLE_TYPE_IBC_WASM_HOOKS_OUTGOING_TRANSFER_TIMEOUT => {
+                    let ibc_lifecycle_complete = serde_json::from_slice::<IBCLifecycleComplete>(&sent_msg.msg);
+                if ibc_lifecycle_complete.is_err(){
+                    trace!("get_verified_msg HANDLE_TYPE_IBC_WASM_HOOKS_OUTGOING_TRANSFER_TIMEOUT: sent_msg.msg cannot be parsed as IBCLifecycleComplete: {:?} Error: {:?}", String::from_utf8_lossy(&sent_msg.msg), ibc_lifecycle_complete.err());
+                    return false;
+                }
+                  let ibc_lifecycle_complete= ibc_lifecycle_complete.unwrap();
+
+              match  ibc_lifecycle_complete {
+                IBCLifecycleComplete::IBCLifecycleComplete(IBCLifecycleCompleteOptions::IBCAck {..}) =>  false                ,
+                IBCLifecycleComplete::IBCLifecycleComplete(IBCLifecycleCompleteOptions::IBCTimeout { channel, sequence  }) =>
+                channel == packet.source_channel
+                && sequence == packet.sequence,
+            }
+                },
+            _ => false,
+
+            }
+        },
```

Comment:
Add functions here and split into multiple files. This is unnecessary bloat

Reply:
Also, you can have some of the repeating stuff be macros to avoid code duplication

Reply:
https://github.com/scrtlabs/SecretNetwork/commit/c3addbc8b